### PR TITLE
user-facing cleanlab.datavaluation module

### PR DESCRIPTION
**Pull request by coding-famer**
## Summary

Solve #973 

> 🎯 **Purpose**: Move the data valuation calculation function to `cleanlab.datavaluation` module and support `features` as input

# Necessary Imports:
`from cleanlab.datavaluation import data_shapley_knn`

## Impact

> 🌐 Areas Affected: `datalab.issue_manager.DataValuationIssueManager`.

**Unaddressed Cases**

One thing may need discussion: when both `knn_graph` and `features` are provided and the given `k` is larger than the `k` in `knn_graph`, the default behavior now is recomputing `knn_graph` using `features.

### CODE REVIEW

File: cleanlab/datalab/internal/issue_manager/data_valuation.py [hunk 1]
```diff
@@ -142,26 +140,22 @@ def find_issues(
...
-                    "metric was specified."
-                )
-            self.metric = knn.metric
-
-            try:
-                check_is_fitted(knn)
-            except NotFittedError:
-                knn.fit(features)
-
-            knn_graph = knn.kneighbors_graph(mode="distance")
```
Comment (elisno) [inline cleanlab/datalab/internal/issue_manager/data_valuation.py:160]: Please revert to this logic on deciding whether to compute a `knn_graph` within Datalab (and this issue manager).
The cleanlab.data_valuation module should not be responsible for computing a knn_graph **and passing it** on to Datalab.
It's fine to copy the helper function to the data_valuation module in the case when features are passed to the function in the `cleanlab.data_valuation`module.

File: cleanlab/datalab/internal/issue_manager/data_valuation.py [hunk 2]
```diff
@@ -142,26 +140,22 @@ def find_issues(
...
+            labels, knn_graph=knn_graph, features=features, metric=self.metric, k=self.k
+        )
+
+        if self.metric and self.metric != new_metric:
+            warnings.warn(
+                f"Metric {self.metric} does not match metric {new_metric} used to fit knn. "
+                "Most likely an existing NearestNeighbors object was passed in, but a different "
+                "metric was specified."
+            )
+        self.metric = new_metric
```
Comment (elisno) [inline cleanlab/datalab/internal/issue_manager/data_valuation.py:158]: See the previous comment about reverting the logic of computing within Datalab.

File: cleanlab/datavaluation.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    if metric == None:
+        metric = "cosine" if features.shape[1] > 3 else "euclidean"
+
+    knn = NearestNeighbors(n_neighbors=k, metric=metric).fit(features)
+    knn_graph = knn.kneighbors_graph(mode="distance")
+    try:
+        check_is_fitted(knn)
+    except NotFittedError:
+        knn.fit(features)
+    return knn_graph, knn.metric
```
Comment (elisno) [inline cleanlab/datavaluation.py:59]: This helper function only needs to compute the `knn_graph` within this module. This module is not responsible for tracking the metric used by the simple function. We're not re-implementing the logic of storing the metric outside of Datalab.

```suggestion
def _process_knn_graph_from_features(features: np.ndarray, metric: Optional[str], k: int = 10) -> csr_matrix:
    """Calculate the knn graph from the features if it is not provided in the kwargs."""
    if k > len(features):  # Ensure number of neighbors less than number of examples
        raise ValueError(
            f"Number of nearest neighbors k={k} cannot exceed the number of examples N={len(features)} passed into the estimator (knn)."
        )

    if metric == None:
        metric = "cosine" if features.shape[1] > 3 else "euclidean"

    knn = NearestNeighbors(n_neighbors=k, metric=metric).fit(features)
    knn_graph = knn.kneighbors_graph(mode="distance")
    try:
        check_is_fitted(knn)
    except NotFittedError:
        knn.fit(features)
    return knn_graph
```

File: cleanlab/datavaluation.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def data_shapley_knn(
+    labels: np.ndarray,
+    metric: str,
+    knn_graph: Optional[csr_matrix] = None,
+    features: Optional[np.ndarray] = None,
+    k: int = 10,
+) -> Tuple[np.ndarray, int, str]:
```
Comment (elisno) [inline cleanlab/datavaluation.py]:
```suggestion
) -> np.ndarray:
```

The original method was only supposed to return the scores ([see comment in issue](https://github.com/cleanlab/cleanlab/issues/973#issuecomment-1922822344)). This public function is not responsible for tracking the other information.

File: cleanlab/datavaluation.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                k = knn_graph.nnz // knn_graph.shape[0]
+                Warning(
+                    f"k is larger than the number of neighbors in the knn graph. Using k={k} instead."
+                )
+        new_metric = metric
+    else:
+        if features is None:
+            raise ValueError("features must be provided if knn_graph is not provided.")
+        knn_graph, new_metric = _process_knn_graph_from_features(features, metric, k)
+    return _knn_shapley_score(knn_graph, labels, k), k, new_metric
```
Comment (elisno) [inline cleanlab/datavaluation.py:100]: The method needs either features or a knn_graph. It's simplest in this function to prefer a knn_graph. Make this explicit in the docstring for this function.

```suggestion
    if knn_graph is None and features is None:
        raise ValueError("Either knn_graph or features must be provided.")
    
    if knn_graph is None:
        knn_graph = _process_knn_graph_from_features(cast(np.ndarray, features), metric, k)
    return _knn_shapley_score(knn_graph, labels, k)
```

**Review by elisno — CHANGES_REQUESTED** [review]
Thanks for the PR.

Remember that the main idea of the PR is to move the `_knn_shapley_score` out of the DataValuationIssueManager and into a `cleanlab.data_valuation` module (which you need to rename correctly, by the way).

To keep the module's scope simple, it can borrow some logic from the Datalab necessary to compute the scores (e.g., how to compute the knn_graph), but it's not supposed to keep track of all this information.

**Review by elisno — APPROVED** [review]
Code LGTM.
Made some docs-related changes.

File: docs/source/index.rst [hunk 6]
```diff
@@ -181,6 +181,7 @@ Link to Cleanlab Studio docs: `help.cleanlab.ai <https://help.cleanlab.ai/>`_
    cleanlab/dataset
    cleanlab/outlier
    cleanlab/multiannotator
+   cleanlab/data_valuation
```
Comment (jwmueller) [inline docs/source/index.rst:184]: put at bottom of the list, not here

File: cleanlab/data_valuation.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+def data_shapley_knn(
+    labels: np.ndarray,
+    knn_graph: Optional[csr_matrix] = None,
+    features: Optional[np.ndarray] = None,
+    metric: Optional[str] = None,
+    k: int = 10,
+) -> np.ndarray:
+    """
+    Compute the Shapley values of data points using a K-Nearest Neighbors (KNN) graph.
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:80]: Would be good to copy some of the text I'd already written describing this method in the issue types guide

Screen Shot 2024-03-25 at 9 34 21 PM

Comment (jwmueller) [reply cleanlab/data_valuation.py:80]: here is where to describe the approximate KNN algorithm, not in the top-level description of this module

File: cleanlab/data_valuation.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+"""
+Provides methods for computing the Shapley values of data points using a K-Nearest Neighbors (KNN) graph.
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:17]: this top-level description of the module should be more vague. Don't mention one specific method or Data Shapely.

Just talk about Data Valuation in general and quantifying the benefits of individual data points to a trained ML model's performance.

That way we can add more methods to the module in the future and don't need to change the top-level description. Also do not need to list out individual methods in the top-level description of a module ever

File: cleanlab/data_valuation.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    Compute the Shapley values of data points using a K-Nearest Neighbors (KNN) graph.
+
+    This function calculates the contribution (Shapley value) of each data point in a dataset
+    for model training, based on the principle that data points contributing more to the
+    model's accuracy are of higher value.
+
+    Parameters
+    ----------
+    labels :
+        An array of labels for the data points.
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:89]: need to mention this is only for multi-class classification datasets

File: cleanlab/data_valuation.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    knn_graph :
+        A precomputed sparse KNN graph. If not provided, it will be computed from the `features` using the specified `metric`.
+    features :
+        The feature matrix of data points. Necessary if `knn_graph` is not supplied.
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:93]: feature matrix of data points is unclear language. Please copy language from elsewhere where we've described features args for eg. Datalab

File: cleanlab/data_valuation.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    knn_graph = knn.kneighbors_graph(mode="distance")
+    try:
+        check_is_fitted(knn)
+    except NotFittedError:
+        knn.fit(features)
+    return knn_graph
+
+
+def data_shapley_knn(
+    labels: np.ndarray,
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:73]:
```suggestion
    labels: np.ndarray, *,
```

Comment (jwmueller) [reply cleanlab/data_valuation.py:73]: we always want to be able to re-order the keyword only args for every method in cleanlab package

File: cleanlab/data_valuation.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    knn_graph: Optional[csr_matrix] = None,
+    features: Optional[np.ndarray] = None,
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:75]: put features arg ahead of knn_graph. Note how all the other args are related to knn_graph. 

Also recommended approach we expect most users to take is to supply features

File: cleanlab/data_valuation.py (hunk 10 above)
Comment (jwmueller) [reply cleanlab/data_valuation.py:93]: swap ordering of features and knn graph in documentation too

File: cleanlab/data_valuation.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    knn_graph :
+        A precomputed sparse KNN graph. If not provided, it will be computed from the `features` using the specified `metric`.
+    features :
+        The feature matrix of data points. Necessary if `knn_graph` is not supplied.
+    metric : Optional[str], default=None
+        The distance metric for KNN graph construction.
+        Supports metrics available in ``sklearn.neighbors.NearestNeighbors``
+        Default metric is ``"cosine"`` for ``dim(features) > 3``, otherwise ``"euclidean"`` for lower-dimensional data.
+    k :
+        The number of neighbors to consider for the KNN graph and Shapley value computation.
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:99]:
```suggestion
        The number of neighbors to consider for the KNN graph and Data Shapley value computation.
```

Comment (jwmueller) [reply cleanlab/data_valuation.py:99]: never say "Shapley value", it's always "Data Shapley value".
The former means something entirely different in ML (Feature attribution)

File: cleanlab/data_valuation.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        Must be less than the total number of data points.
+        The value may not exceed the number of neighbors of each data point stored in the KNN graph.
+
+    Returns
+    -------
+    scores :
+        An array containing transformed Shapley values for each data point, calibrated to indicate their relative importance.
+        These scores have been adjusted to fall within the range of 0 to 1.
+        Values closer to 1 suggest data points are highly influential and positively contribute to the model's performance.
+        Conversely, scores below 0.5 are interpreted as indicating a negative impact on model performance.
```
Comment (jwmueller) [inline cleanlab/data_valuation.py:109]:
```suggestion
    scores :
        An array of transformed Data Shapley values for each data point, calibrated to indicate their relative importance.
        These scores have been adjusted to fall within 0 to 1.
        Values closer to 1 indicate data points that are highly influential and positively contribute to a trained ML model's performance.
        Conversely, scores below 0.5 indicate data points estimated to  negatively impact model performance.
```

File: cleanlab/data_valuation.py (hunk 7 above)
Comment (coding-famer) [reply cleanlab/data_valuation.py]: BTW, the document in this figure seems to need updating too (data valuation issue now can accept features as input), can I update it in this PR?

Comment (jwmueller) [reply cleanlab/data_valuation.py]: yes please! Thanks!

**Review by jwmueller — APPROVED** [review]
lgtm AFAICT, leaving merge to @elisno

**Review by elisno — APPROVED** [review]
